src/Runner.Worker/JobRunner.cs

```diff
@@ -206,6 +217,59 @@ public async Task<TaskResult> RunAsync(Pipelines.AgentJobRequestMessage message,
             }
         }
 
+        private async Task<TaskResult> CompleteJobAsync(IRunServer runServer, IJobServer jobServer, IExecutionContext jobContext, Pipelines.AgentJobRequestMessage message, TaskResult? taskResult = null)
+        {
+            if (runServer != null)
+            {
+                return await CompleteJobAsync(runServer, jobContext, message, taskResult);
+            }
+            else
+            {
+                return await CompleteJobAsync(jobServer, jobContext, message, taskResult);
+            }
+        }
+
+        private async Task<TaskResult> CompleteJobAsync(IRunServer runServer, IExecutionContext jobContext, Pipelines.AgentJobRequestMessage message, TaskResult? taskResult = null)
+        {
+            jobContext.Debug($"Finishing: {message.JobDisplayName}");
+            TaskResult result = jobContext.Complete(taskResult);
+            if (jobContext.Global.Variables.TryGetValue("Node12ActionsWarnings", out var node12Warnings))
+            {
+                var actions = string.Join(", ", StringUtil.ConvertFromJson<HashSet<string>>(node12Warnings));
+                jobContext.Warning(string.Format(Constants.Runner.Node12DetectedAfterEndOfLife, actions));
+            }
+
+            // Load any upgrade telemetry
+            LoadFromTelemetryFile(jobContext.Global.JobTelemetry);
+
+            // Make sure we don't submit secrets as telemetry
+            MaskTelemetrySecrets(jobContext.Global.JobTelemetry);
+
+            Trace.Info($"Raising job completed");
```

Comment:
```suggestion
            Trace.Info($"Raising job completed against run service");
```

Just for my own sanity if we could 🥇